flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/UnionInputGate.java

```diff
@@ -61,7 +61,7 @@
  * +--------------------+
  * </pre>
  *
- * It is possible to recursively union union input gates.
+ * <strong>It is NOT possible to recursively union union input gates.</strong>
```

Comment:
I was thinking whether this class should take `SingleInputGate`s in the constructor instead of `InputGate`s and thus we could remove `pollNextBufferOrEvent` method from the interface and move it to the `SingleInputGate` only.

Reply:
That would be a more invasive change in several places using `InputGate`s and we actually are able to use any other input gate types (if there were some) - just the recursive union does not make sense. So I guess, excluding ourself is the better approach here in terms of extensibility. Think of it as we're excluding this use not because we did not implement `pollNextBuffer` but because it does not make sense and we do not want this recursion.